src/result.ts

```diff
@@ -18,13 +18,54 @@ import {
   PredicateOfT,
   ResultMatcher,
   ResultMatcherNoReturn,
+  ResultValueOf,
   Some,
 } from './utilities';
 
 /**
  * Represents a successful or failed operation
  */
 export class Result<TValue = Unit, TError = string> {
+  /**
+   * Combines several results (and any error messages) into a single result.
+   * The returned result will be a failure if any of the input results are failures.
+   *
+   * @param results The Results to be combined.
+   * @returns A Result that is a success when all the input results are also successes.
+   */
+  static combine<T extends Record<string, Result<unknown>>>(
+    results: T
+  ): Result<{ [K in keyof T]: ResultValueOf<T[K]> }> {
+    const resultEntries = Object.entries(results);
+
+    const failedResults = resultEntries.filter(
+      ([, result]) => result.isFailure
+    );
+    const succeededResults = resultEntries.filter(
+      ([, result]) => result.isSuccess
+    );
+
+    if (failedResults.length === 0) {
+      const values = succeededResults.reduce(
+        (resultValues, [key, result]) => ({
+          ...resultValues,
```

Comment:
I'm not sure how the JS runtime could optimize this 🤔

It won't matter for small numbers of results, but continuously creating new objects by destructuring all the properties of previous ones in a loop wouldn't be great for performance.

It could be avoided by keeping the same object an assigning new properties/values to it.

```ts
.reduce((resultValues, [key, result]) => {
   resultValues[key] = result.getValueOrThrow();
   
   return resultValues;
}, {});
```
It's not as nice looking 😎 as your expression based approach.